### PR TITLE
add Keycloak provider

### PR DESCRIPTION
**Pull request by bbigras**
## Reasoning 💡

Add a Keycloak provider.

I tested with a self-hosted Keycloak.

I based this PR on the auth0 provider.

I was first wondering if it was possible in https://github.com/nextauthjs/next-auth/pull/2411#issuecomment-892768889.

## Checklist 🧢

- [x] Documentation
- [ ] Tests
- [x] Ready to be merged

## Affected issues 🎟

Related: #1261

### CODE REVIEW

**Comment by balazsorban44** [comment]
Awesome, so I assume you were able to get it to work! 🎉 I will have a closer look. From now on, I will want us to enforce stricter defaults for built-in providers for consistency, and I also want to see if we can get an account on the provider platform for testing purposes before merging. But this looks promising.

**Comment by bbigras** [comment]
> so I assume you were able to get it to work!

Yep, and I just tested with please-open.it.

To reproduce:

- create client with `http://localhost:3000` as the root URL
- set `Access Type` to `confidential` (the secret will be in the `credentials` tab after save)
- you can create a user, but when I tested it used mine as I was already logged.

`KEYCLOAK_ISSUER=https://app.please-open.it/auth/realms/<realm_id>`

**Review by chuongxl — APPROVED** [review]

**Comment by masone** [comment]
Apart from testing, is there something we can do to help moving this forward?

**Comment by balazsorban44** [comment]
I'm just want to have  a look myself before merging. Just use the source code as a custom provider if you need this already. Currently busy with migrating our adapters.

**Comment by bbigras** [comment]
> There is also a merge conflict. @bbigras could you address that please?

@balazsorban44 Done. It seems a rebase was enough.

File: src/providers/keycloak.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    wellKnown: `${options.issuer}/.well-known/openid-configuration`,
+    type: "oauth",
+    authorization: { params: { scope: "openid email profile" } },
+    checks: ["pkce", "state"],
+    idToken: true,
+    profile(profile) {
+      return {
+        id: profile.sub,
+        name: profile.name,
+        email: profile.email,
```
Comment (balazsorban44) [inline src/providers/keycloak.js:15]:
```suggestion
        name: profile.name ?? profile.preferred_username,
        email: profile.email,
        image: null
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Awesome, finally got around to check this out! :+1: Looks good, I only had a small comment/suggestion. Other than that, it is good to go!

**Comment by bbigras** [comment]
Done. Do you want me to squash it, or do you want to do it while merging or something?

**Comment by dicksonmelo** [comment]
> It's merged! Give me a sec, and I'll release it too.
> 
> Here it is: `v4.0.0-next.24`

Is there any forecast when it can be used?

**Comment by dicksonmelo** [comment]
> It's merged! Give me a sec, and I'll release it too.
> 
> Here it is: `v4.0.0-next.24`

@balazsorban44 
Is there any forecast when it can be used? Can I download this v4.0.0-next.24 release?

**Comment by balazsorban44** [comment]
Absolutely, `npm install next-auth@4.0.0-next.24` 😉. Make sure you don't have a caret (^) in the version number in package.json. Had a small trouble with NPM a while ago and there is a buggy 4.0.0 version out unfortunately. 😬 So you have to force the exact version above. This will be solved when we release this into stable.

**Comment by dicksonmelo** [comment]
I downloaded the version, but I was trying to use it like the previous versions. Thank you so much!

**Pull request merged**